src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs

```diff
@@ -1,184 +1,134 @@
+
+            // Forbid codepoints which aren't mapped to characters or which are otherwise always disallowed
+            // (includes categories Cc, Cs, Co, Cn, Zs [except U+0020 SPACE], Zl, Zp)
+            _allowedCharsBitmap.ForbidUndefinedCharacters();
+
+            // Forbid characters that are special in HTML.
+            // Even though this is a common encoder used by everybody (including URL
+            // and JavaScript strings), it's unfortunately common for developers to
+            // forget to HTML-encode a string once it has been URL-encoded or
+            // JavaScript string-escaped, so this offers extra protection.
+            _allowedCharsBitmap.ForbidCharacter('<');
+            _allowedCharsBitmap.ForbidCharacter('>');
+            _allowedCharsBitmap.ForbidCharacter('&');
+            _allowedCharsBitmap.ForbidCharacter('\''); // can be used to escape attributes
+            _allowedCharsBitmap.ForbidCharacter('\"'); // can be used to escape attributes
+            _allowedCharsBitmap.ForbidCharacter('+'); // technically not HTML-specific, but can be used to perform UTF7-based attacks
         }
 
-        [MethodImpl(MethodImplOptions.NoInlining)] // the JITter can attempt to inline the caller itself without worrying about us
-        private static HtmlEncoder CreateDefaultEncoderSlow()
+        public DefaultHtmlEncoder(params UnicodeRange[] allowedRanges) : this(new CodePointFilter(allowedRanges))
+        { }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public override bool Encodes(int unicodeScalar)
         {
-            var onDemandEncoder = new HtmlEncoder();
-            return Interlocked.CompareExchange(ref _defaultEncoder, onDemandEncoder, null) ?? onDemandEncoder;
+            if (UnicodeHelpers.IsSupplementaryCodePoint(unicodeScalar)) return true;
+            return !_allowedCharsBitmap.IsUnicodeScalarAllowed(unicodeScalar);
         }
 
-        /// <summary>
-        /// Everybody's favorite HtmlEncode routine.
-        /// </summary>
-        public void HtmlEncode(char[] value, int startIndex, int charCount, TextWriter output)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe override int FindFirstCharacterToEncode(char* text, int charCount)
         {
-            _innerUnicodeEncoder.Encode(value, startIndex, charCount, output);
+            for (int i = 0; i < charCount; i++)
+            {
+                var character = *text;
+                if (!_allowedCharsBitmap.IsCharacterAllowed(character)) { return i; }
+                text++;
+            }
+            return -1;
         }
 
-        /// <summary>
-        /// Everybody's favorite HtmlEncode routine.
-        /// </summary>
-        public string HtmlEncode(string value)
+        public override int MaxOutputCharsPerInputChar
         {
-            return _innerUnicodeEncoder.Encode(value);
+            get { return 8; }
```

Comment:
Could you either move this to a named constant or add a comment explaining where this number comes from?